File: Cpp/LinkedListBenchmark.cpp
#include <benchmark/benchmark.h>

#include "LinkedList.hpp"

bool is_odd(long long n) { return n % 2 == 1; }
long long sum(long long a, long long b) { return a + b; }

LinkedList<long long> createList(int size) {
  auto list = LinkedList<long long>();

  for (int i = 0; i < size; i++) {
    list.push(i + 1);
  }

  return list;
}

static void BM_SortList(benchmark::State& state) {
  auto size = state.range(0);
  auto list = createList(size);

  for (auto _ : state) list.mergeSort();
}

static void BM_FilterList(benchmark::State& state) {
  auto size = state.range(0);
  auto list = createList(size);
  for (auto _ : state) list.filter(is_odd);
}

static void BM_ReduceList(benchmark::State& state) {
  auto size = state.range(0);
  auto list = createList(size);
  for (auto _ : state) list.reduce(sum, 0);
}

BENCHMARK(BM_SortList)
    ->Arg(10)
    ->Arg(100)
    ->Arg(1000)
    ->Arg(10000)
    ->Arg(100000)
    ->Arg(250000)
    ->Arg(500000)
    ->Arg(750000)
    ->Arg(1000000);

BENCHMARK(BM_FilterList)
    ->Arg(10)
    ->Arg(100)
    ->Arg(1000)
    ->Arg(10000)
    ->Arg(100000)
    ->Arg(250000)
    ->Arg(500000)
    ->Arg(750000)
    ->Arg(1000000)
    ->Iterations(1);

BENCHMARK(BM_ReduceList)
    ->Arg(10)
    ->Arg(100)
    ->Arg(1000)
    ->Arg(10000)
    ->Arg(100000)
    ->Arg(250000)
    ->Arg(500000)
    ->Arg(750000)
    ->Arg(1000000);

BENCHMARK_MAIN();

File: TypeScript/LinkedListBenchmark.ts
import {LinkedList} from './LinkedList';
const { add, cycle, suite, complete, save } = require('benny');

function createList(size: number): LinkedList<number> {
  let list = new LinkedList<number>();

  for (let i = 0; i < size; i++) {
    list.push(i + 1);
  }

  return list;
}

function isOdd(n: number): boolean {
  return n % 2 === 1;
}

function sum(a: number, b: number): number {
  return a + b;
}

const sortSuite = (...sizes: number[]) => {
  return sizes.map((size) => {
    return add(`${size} elements`, () => {
      const input = createList(size)

      return () => input.mergeSort()
    });
  });
}

const filterSuite = (...sizes: number[]) => {
  return sizes.map((size) => {
    return add(`${size} elements`, () => {
      const input = createList(size)

      return () => input.filter(isOdd)
    });
  });
}

const reduceSuite = (...sizes: number[]) => {
  return sizes.map((size) => {
    return add(`${size} elements`, () => {
      const input = createList(size)

      return () => input.reduce(sum, 0)
    });
  });
}

const batches = [10, 100, 500, 1000, 2500, 5000, 7500, 10000];

suite(
  'Sort',
  ...sortSuite(...batches),
  cycle(),
  complete(),
  save({file: 'LL_sort'})
)

suite(
  'Filter',
  ...filterSuite(...batches),
  cycle(),
  complete(),
  save({file: 'LL_filter'})
)

suite(
  'Reduce',
  ...reduceSuite(...batches),
  cycle(),
  complete(),
  save({file: 'LL_reduce'})
)
